File: submission/example-bundle/evaluation/adapter.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

PLAN = Path("/evaluation/input/plan.json")
CAPTURED = Path("/evaluation/trusted/executions.json")
ROOT = Path("/workspace/repo")
CANARY = "TB_HIDDEN_CANARY_6F2F5A64B88D"


def _selectors(plan: dict[str, Any]) -> list[str]:
    return [
        *(item["selector"] for item in plan["pass_to_pass"]),
        *(item["selector"] for item in plan["fail_to_pass"]),
    ]


def _build_plan() -> int:
    plan = json.loads(PLAN.read_text(encoding="utf-8"))
    executions = [
        {
            "execution_id": f"selector-{index:03d}",
            "requested_selectors": [selector],
            "argv": [
                "python",
                "/evaluation/harness/adapter.py",
                "run-selector",
                selector,
            ],
            "timeout_seconds": plan["timeout_seconds"],
        }
        for index, selector in enumerate(_selectors(plan), start=1)
    ]
    print(json.dumps({"schema_version": "2", "executions": executions}))
    return 0


def _run_selector(selector: str) -> int:
    if (ROOT / ".task_bundle_hidden_canary").read_text().strip() != CANARY:
        return 2
    statuses = {
        "synthetic::readme": (ROOT / "README").is_file(),
        "synthetic::answer": (
            (ROOT / "answer.txt").is_file()
            and (ROOT / "answer.txt").read_bytes() == b"42\n"
        ),
    }
    return 0 if statuses.get(selector, False) else 1


def _parse_result() -> int:
    captured = json.loads(CAPTURED.read_text(encoding="utf-8"))["executions"]
    tests: list[dict[str, Any]] = []
    collection_succeeded = True
    for execution in captured:
        selectors = execution["requested_selectors"]
        if len(selectors) != 1:
            raise ValueError("synthetic adapter requires one selector per execution")
        if execution["stdout_truncated"] or execution["stderr_truncated"]:
            raise ValueError("captured test output was truncated")
        selector = selectors[0]
        if execution["timed_out"]:
            status = "timeout"
        elif execution["exit_code"] == 0:
            status = "passed"
        elif execution["exit_code"] == 1:
            status = "failed"
        else:
            status = "error"
            collection_succeeded = False
        tests.append(
            {
                "requested_selector": selector,
                "observed_id": selector,
                "status": status,
                "duration_ms": execution["duration_ms"],
                "message": (execution["stderr"] or execution["stdout"])[:16_384]
                or None,
            }
        )
    result = {
        "schema_version": "1",
        "framework": "synthetic-python-exit-code",
        "harness_status": (
            "completed" if collection_succeeded else "collection_failed"
        ),
        "collection_succeeded": collection_succeeded,
        "execution_started": bool(captured),
        "command": ["python", "/evaluation/harness/adapter.py", "run-selector"],
        "started_at": captured[0]["started_at"],
        "finished_at": captured[-1]["finished_at"],
        "exit_code": max(
            (
                execution["exit_code"]
                for execution in captured
                if execution["exit_code"] is not None
            ),
            default=0,
        ),
        "tests": tests,
    }
    print(json.dumps(result, separators=(",", ":")))
    return 0


def main() -> int:
    if len(sys.argv) < 2:
        return 2
    if sys.argv[1] == "build-plan":
        return _build_plan()
    if sys.argv[1] == "run-selector" and len(sys.argv) == 3:
        return _run_selector(sys.argv[2])
    if sys.argv[1] == "parse-result":
        return _parse_result()
    return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: bundles/synthetic-go-calculator/evaluation/adapter.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"time"
)

type selector struct {
	Selector string `json:"selector"`
}

type plan struct {
	PassToPass     []selector `json:"pass_to_pass"`
	FailToPass     []selector `json:"fail_to_pass"`
	TimeoutSeconds int        `json:"timeout_seconds"`
}

type executionPlanItem struct {
	ExecutionID        string   `json:"execution_id"`
	RequestedSelectors []string `json:"requested_selectors"`
	Argv               []string `json:"argv"`
	TimeoutSeconds     int      `json:"timeout_seconds"`
}

type executionPlan struct {
	SchemaVersion string              `json:"schema_version"`
	Executions    []executionPlanItem `json:"executions"`
}

type capturedExecution struct {
	ExecutionID        string   `json:"execution_id"`
	RequestedSelectors []string `json:"requested_selectors"`
	Argv               []string `json:"argv"`
	StartedAt          string   `json:"started_at"`
	FinishedAt         string   `json:"finished_at"`
	DurationMS         int      `json:"duration_ms"`
	ExitCode           *int     `json:"exit_code"`
	TimedOut           bool     `json:"timed_out"`
	Stdout             string   `json:"stdout"`
	Stderr             string   `json:"stderr"`
	StdoutTruncated    bool     `json:"stdout_truncated"`
	StderrTruncated    bool     `json:"stderr_truncated"`
}

type capturedSet struct {
	Executions []capturedExecution `json:"executions"`
}

type event struct {
	Action  string  `json:"Action"`
	Test    string  `json:"Test"`
	Elapsed float64 `json:"Elapsed"`
}

type testResult struct {
	RequestedSelector string `json:"requested_selector"`
	ObservedID        string `json:"observed_id"`
	Status            string `json:"status"`
	DurationMS        int    `json:"duration_ms"`
}

type result struct {
	SchemaVersion       string       `json:"schema_version"`
	Framework           string       `json:"framework"`
	HarnessStatus       string       `json:"harness_status"`
	CollectionSucceeded bool         `json:"collection_succeeded"`
	ExecutionStarted    bool         `json:"execution_started"`
	Command             []string     `json:"command"`
	StartedAt           string       `json:"started_at"`
	FinishedAt          string       `json:"finished_at"`
	ExitCode            int          `json:"exit_code"`
	Tests               []testResult `json:"tests"`
}

func requestedSelectors(value plan) []selector {
	return append(value.PassToPass, value.FailToPass...)
}

func buildPlan() error {
	planBytes, err := os.ReadFile("/evaluation/input/plan.json")
	if err != nil {
		return err
	}
	var requested plan
	if err := json.Unmarshal(planBytes, &requested); err != nil {
		return err
	}
	executions := []executionPlanItem{}
	for index, wanted := range requestedSelectors(requested) {
		executions = append(executions, executionPlanItem{
			ExecutionID:        fmt.Sprintf("selector-%03d", index+1),
			RequestedSelectors: []string{wanted.Selector},
			Argv: []string{
				"/usr/bin/env",
				"HOME=/workspace/task-bundle-home",
				"GOCACHE=/workspace/task-bundle-go-cache",
				"GOTMPDIR=/workspace/task-bundle-go-tmp",
				"go", "test", "-json", "-run",
				"^" + regexp.QuoteMeta(wanted.Selector) + "$", ".",
			},
			TimeoutSeconds: requested.TimeoutSeconds,
		})
	}
	return json.NewEncoder(os.Stdout).Encode(executionPlan{
		SchemaVersion: "2",
		Executions:    executions,
	})
}

func parseResult() error {
	payload, err := os.ReadFile("/evaluation/trusted/executions.json")
	if err != nil {
		return err
	}
	var captured capturedSet
	if err := json.Unmarshal(payload, &captured); err != nil {
		return err
	}
	tests := []testResult{}
	collectionSucceeded := true
	maximumExit := 0
	for _, execution := range captured.Executions {
		if len(execution.RequestedSelectors) != 1 {
			return fmt.Errorf("synthetic adapter requires one selector per execution")
		}
		requestedSelector := execution.RequestedSelectors[0]
		status := "error"
		elapsed := execution.DurationMS
		observedID := requestedSelector
		if execution.StdoutTruncated || execution.StderrTruncated {
			return fmt.Errorf("captured test output was truncated")
		}
		if execution.TimedOut {
			status = "timeout"
		} else {
			observed := map[string]event{}
			scanner := bufio.NewScanner(bytes.NewReader([]byte(execution.Stdout)))
			for scanner.Scan() {
				var item event
				if json.Unmarshal(scanner.Bytes(), &item) == nil && item.Test != "" {
					if item.Action == "pass" || item.Action == "fail" ||
						item.Action == "skip" {
						observed[item.Test] = item
					}
				}
			}
			if err := scanner.Err(); err != nil {
				return err
			}
			item, ok := observed[requestedSelector]
			if !ok {
				status = "missing"
				collectionSucceeded = false
				observedID = ""
			} else {
				statuses := map[string]string{
					"pass": "passed",
					"fail": "failed",
					"skip": "skipped",
				}
				status = statuses[item.Action]
				elapsed = int(item.Elapsed * 1000)
			}
		}
		if execution.ExitCode != nil && *execution.ExitCode > maximumExit {
			maximumExit = *execution.ExitCode
		}
		if execution.ExitCode != nil && *execution.ExitCode > 1 {
			collectionSucceeded = false
		}
		tests = append(tests, testResult{
			RequestedSelector: requestedSelector,
			ObservedID:        observedID,
			Status:            status,
			DurationMS:        elapsed,
		})
	}
	started := time.Now().UTC().Format(time.RFC3339Nano)
	finished := started
	if len(captured.Executions) > 0 {
		started = captured.Executions[0].StartedAt
		finished = captured.Executions[len(captured.Executions)-1].FinishedAt
	}
	output := result{
		SchemaVersion:       "1",
		Framework:           "go-test-json",
		HarnessStatus:       "completed",
		CollectionSucceeded: collectionSucceeded,
		ExecutionStarted:    len(captured.Executions) > 0,
		Command:             []string{"go", "test", "-json", "-run", "<selector>", "."},
		StartedAt:           started,
		FinishedAt:          finished,
		ExitCode:            maximumExit,
		Tests:               tests,
	}
	if !collectionSucceeded {
		output.HarnessStatus = "collection_failed"
	}
	return json.NewEncoder(os.Stdout).Encode(output)
}

func main() {
	var err error
	if len(os.Args) != 2 {
		err = fmt.Errorf("usage: adapter build-plan|parse-result")
	} else if os.Args[1] == "build-plan" {
		err = buildPlan()
	} else if os.Args[1] == "parse-result" {
		err = parseResult()
	} else {
		err = fmt.Errorf("unknown adapter mode")
	}
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(2)
	}
}
